The Radon-Nikodym theorem in Lean blog (#2)
leanprover-community/blog: posts/the-radon-nikodym-theorem-in-lean.md

```
Since in order to prove the Lebesgue decomposition for $\sigma$-finite measures, 
we will first need to show it for finite measures, this definition allows us 
to reuse the same statement for both cases, avoiding duplicate code. Furthermore, 
as we would like to extract the measure and measurable function from the decomposition, 
we may define functions that chooses the decomposition if it exists, and 
zero otherwise.
```lean
def singular_part (μ ν : measure α) : measure α :=
if h : have_lebesgue_decomposition μ ν then (classical.some h.lebesgue_decomposition).1 else 0

def radon_nikodym_deriv (μ ν : measure α) : α → ℝ≥0∞ :=
if h : have_lebesgue_decomposition μ ν then (classical.some h.lebesgue_decomposition).2 else 0
```
With the Lebesgue decomposition for $\sigma$-finite measures formalized, the 
Radon-Nikodym theorem follows easily. Furthermore, the generalisation of the 
Radon-Nikodym theorem to signed measures follows also, simply by utilising 
the Jordan decomposition and realising the Radon-Nikodym derivative of 
the signed measure is the difference of the Radon-Nikodym derivative of 
the parts of the Jordan decomposition. However, using this definition of the 

> Radon-Nikodym derivative for the signed measures possess a problem, in which 

Reviewer: ```suggestion
Radon-Nikodym derivative for the signed measures poses a problem, in which 
```

proving any properties about them requires an absolutely continuous condition.
Thus, it was decided to generalise the Lebesgue decomposition theorem for 
signed measures from which we obtain the general Radon-Nikodym theorem.

Similar to the Lebesgue decomposition for the positive measures, the Lebesgue 
decomposition between a signed measure and a positive measure states that, 
given a signed measure $s$ and a $\sigma$-finite measure $\mu$, there exists 
a measurable function $f : \alpha \to \mathbb{R}$ and a signed measure $t$, such 
that $t \perp \mu$ and $s = t + f\mu$.
While this version of the Lebesgue decomposition was also represented as a class, 
the statement itself was modified to be an equivalent, yet easier to work with 
statement. Namely, a signed measure $s$ has Lebesgue decomposition with respect to 
a measure $\mu$ if both parts of the Jordan decomposition of $s$ have Lebesgue 
decomposition to $\mu$.
```lean

> class have_lebesgue_decomposition (s : signed_measure α) (μ : measure α) : Prop :=

Reviewer: It might make sense to give a more qualified name here (is this `signed_measure.have_lebesgue_decomposition`?) so as to distinguish this from the other `have_lebesgue_decomposition` above.

description: ''
has_math: true
link: ''
slug: the-radon-nikodym-theorem-in-lean
tags: ''
title: The Radon-Nikodym theorem in Lean
type: text
---

I have for the past two months been working on formalising the Radon-Nikodym theorem 
in Lean, and with [PR #9065](https://github.com/leanprover-community/mathlib/pull/9065) 
merged into mathlib, this journey seems to have finally come to an end. 

The Radon-Nikodym theorem provides a necessary and sufficient condition for 
comparing two measures, and allows us (under certain conditions) to express 
one measure in terms of another.
The Radon-Nikodym theorem is an important result in measure theory and has a
wide range of applications in different fields of mathematics. Most notably, 
it can be applied in probability theory in the definition 
of the conditional expectation and in mathematical finance through the Girsanov 
theorem[^1].

[^1]: [wikipedia.org/wiki/Radon-Nikodym_theorem#Applications](https://en.wikipedia.org/wiki/Radon%E2%80%93Nikodym_theorem#Applications)

Given two measures $\mu$ and $\nu$ on $\alpha$, we say that $\mu$ is 
*absolutely continuous* with respect to $\nu$ (and write $\mu \ll \nu$) if for all 

> $S \subseteq \alpha$, $\nu(S) = 0$ implies $\mu(S) = 0$. Absolutely continuity 

Reviewer 2: ```suggestion
subset $S$ of $\alpha$, $\nu(S) = 0$ implies $\mu(S) = 0$. Absolute continuity 
```
The first edit tries to soften the discrepancy between informal math pretending to be based on set theory and Lean code.

is an important notion for measures and we would like to establish a condition 
for when it is true. 

Given a measure $\mu$ on $\alpha$ and a measurable function 
$f : \alpha \to \overline{\mathbb{R}}_{\ge 0}$, the set function 
$$S \mapsto \int_S f \text{ d} \mu$$
is also a measure on $\alpha$ and we denote this measure by $f\mu$. It is 
easy, and intuitive to see that $f\mu \ll \mu$, however, it is not clear 
whether the reverse is true. The Radon-Nikodym theorem proves the reverse 
implication for certain measures. In particular, the classical Radon-Nikodym 
theorem states that, two $\sigma$-finite measures $\mu, \nu$ satisfies $\mu \ll \nu$ 
if and only if there exists a *essentially unique* measurable function $f$ 
such that $\mu = f\nu$. This function is known as the Radon-Nikodym derivative 
and is denoted by $\frac{\text{d}\mu}{\text{d}\nu}$.

In Lean, this is shown by 
```lean
theorem absolutely_continuous_iff_with_density_radon_nikodym_deriv_eq
  {μ ν : measure α} [have_lebesgue_decomposition μ ν] :
  μ ≪ ν ↔ ν.with_density (radon_nikodym_deriv μ ν) = μ

> ```

Reviewer 2: This is very hard to related to the above discussion because of the `have_lebesgue_decomposition μ ν` assumption which doesn't appear at all above.

Author: I've moved the lean code after the section about `have_lebesgue_decomposition` so perhaps it makes more sense

wide range of applications in different fields of mathematics. Most notably, 
it can be applied in probability theory in the definition 
of the conditional expectation and in mathematical finance through the Girsanov 
theorem[^1].

[^1]: [wikipedia.org/wiki/Radon-Nikodym_theorem#Applications](https://en.wikipedia.org/wiki/Radon%E2%80%93Nikodym_theorem#Applications)

Given two measures $\mu$ and $\nu$ on $\alpha$, we say that $\mu$ is 
*absolutely continuous* with respect to $\nu$ (and write $\mu \ll \nu$) if for all 
subsets $S$ of $\alpha$, $\nu(S) = 0$ implies $\mu(S) = 0$. Absolute continuity 
is an important notion for measures and we would like to establish a condition 
for when it is true. 

Given a measure $\mu$ on $\alpha$ and a measurable function 
$f : \alpha \to \overline{\mathbb{R}}_{\ge 0}$, the set function 
$$S \mapsto \int_S f \text{ d} \mu$$
is also a measure on $\alpha$ and we denote this measure by $f\mu$. It is 
easy, and intuitive to see that $f\mu \ll \mu$, however, it is not clear 
whether the reverse is true. The Radon-Nikodym theorem proves the reverse 
implication for certain measures. In particular, the classical Radon-Nikodym 

> theorem states that, two $\sigma$-finite measures $\mu, \nu$ satisfies $\mu \ll \nu$ 

Reviewer 3: ```suggestion
theorem states that two $\sigma$-finite measures $\mu, \nu$ satisfy $\mu \ll \nu$ 
```

if and only if there exists a *essentially unique* measurable function $f$ 

> such that $\mu = f\nu$. This function is known as the Radon-Nikodym derivative 

Reviewer 3: Would it be clearer here to just say "if and only if there exists a measurable function ...", and then in a second sentence "When such a function exists it is essentially unique."?

decomposition states that given two $\sigma$-finite measures $\mu$ and $\nu$, 
there exists a measurable function $f : \alpha \to \overline{\mathbb{R}}_{\ge 0}$ 
and a finite measure $\xi$ such that $\xi$ is mutually singular with respect 
to $\nu$ (denoted $\xi \perp \nu$) and $\mu = \xi + f\nu$.
As with the Jordan decomposition theorem, it was not clear how to represent this 
statement. In particular, it is important for us to be able to extract the aforementioned 
$f$ from the decomposition as this is the Radon-Nikodym derivative. After different
experimentations, the statement was in the end decided to be represented as 
a class. 
```lean
class measure.have_lebesgue_decomposition (μ ν : measure α) : Prop :=
(lebesgue_decomposition :
  ∃ (p : measure α × (α → ℝ≥0∞)), measurable p.2 ∧ p.1 ⊥ₘ ν ∧ μ = p.1 + ν.with_density p.2)
```
Since in order to prove the Lebesgue decomposition for $\sigma$-finite measures, 
we will first need to show it for finite measures, this definition allows us 
to reuse the same statement for both cases, avoiding duplicate code. Furthermore, 
as we would like to extract the measure and measurable function from the decomposition, 

> we may define functions that chooses the decomposition if it exists, and 
> zero otherwise.

Reviewer 3: ```suggestion
we may define a functions that choose the decomposition if it exists, and are
zero otherwise.
```

```lean
def measure.singular_part (μ ν : measure α) : measure α :=
if h : have_lebesgue_decomposition μ ν then (classical.some h.lebesgue_decomposition).1 else 0

def measure.radon_nikodym_deriv (μ ν : measure α) : α → ℝ≥0∞ :=
if h : have_lebesgue_decomposition μ ν then (classical.some h.lebesgue_decomposition).2 else 0
```
With the Lebesgue decomposition for $\sigma$-finite measures formalized, the 

> Radon-Nikodym theorem follows easily by condiering that the singular part of the 

Reviewer 3: ```suggestion
Radon-Nikodym theorem follows easily by considering that the singular part of the 
```

Lebesgue decomposition is zero in the case that $\mu \ll \nu$.
```lean
theorem absolutely_continuous_iff_with_density_radon_nikodym_deriv_eq
  {μ ν : measure α} [have_lebesgue_decomposition μ ν] :
  μ ≪ ν ↔ ν.with_density (radon_nikodym_deriv μ ν) = μ
```
Furthermore, the generalisation of the 

> Radon-Nikodym theorem to signed measures follows also, simply by utilising 

Reviewer 3: Perhaps "furthermore" and "also" are redundant?

the Jordan decomposition and realising the Radon-Nikodym derivative of 

> the signed measure is the difference of the Radon-Nikodym derivative of 

Reviewer 3: ```suggestion
the signed measure is the difference of the Radon-Nikodym derivatives of 
```

Given a measure $\mu$ on $\alpha$ and a measurable function 
$f : \alpha \to \overline{\mathbb{R}}_{\ge 0}$, the set function 
$$S \mapsto \int_S f \text{ d} \mu$$
is also a measure on $\alpha$ and we denote this measure by $f\mu$. It is 
easy, and intuitive to see that $f\mu \ll \mu$, however, it is not clear 
whether the reverse is true. The Radon-Nikodym theorem proves the reverse 
implication for certain measures. In particular, the classical Radon-Nikodym 
theorem states that two $\sigma$-finite measures $\mu, \nu$ satisfy $\mu \ll \nu$ 
if and only if there exists a measurable function $f$ such that $\mu = f\nu$. 
This function is known as the Radon-Nikodym derivative 
(denoted by $\frac{\text{d}\mu}{\text{d}\nu}$) and is *essentially unique** whenever 
it exists.

In Lean, this is shown by `absolutely_continuous_iff_with_density_radon_nikodym_deriv_eq`
and can be found in `measure_theory/decomposition/radon_nikodym`.
While Radon-Nikodym is the main motivation, the proof of the theorem 
itself is rather simple once we have the prerequisites, and thus, the project 
spanned over multiple files, most of which can be found in `measure_theory/decomposition`.

> Of all the prerequisites, the most important are are the 

Reviewer 3: ```suggestion
Of all the prerequisites, the most important are the 
```

as we would often like to transport between signed measures and Jordan 
decompositions, it is easier to work with types rather than propositions since 
this required a construction of an equivalence between the two types. 

[^2]: Given a signed measure `s`, a measure `μ` and a real number `r`, one may show 
`(r • s).singular_part μ = r • s.singular_part μ` without 
requiring `have_lebesgue_decomposition s μ` by proving that the scalar product 
with a real number and the equivalence between signed measures and Jordan 
decompositions commute. 

A similar situation was reached with the Lebesgue decomposition. The Lebesgue 
decomposition states that given two $\sigma$-finite measures $\mu$ and $\nu$, 
there exists a measurable function $f : \alpha \to \overline{\mathbb{R}}_{\ge 0}$ 
and a finite measure $\xi$ such that $\xi$ is mutually singular with respect 
to $\nu$ (denoted $\xi \perp \nu$) and $\mu = \xi + f\nu$.
As with the Jordan decomposition theorem, it was not clear how to represent this 
statement. In particular, it is important for us to be able to extract the aforementioned 
$f$ from the decomposition as this is the Radon-Nikodym derivative. After different

> experimentations, the statement was in the end decided to be represented as 

Reviewer 3: Hopefully this is helpful rather than nitpicking, but: this sentence is unnecessarily convoluted and indirect. How about "After some experiments, I decided to represent this condition using a class."?

a class. 
```lean
class measure.have_lebesgue_decomposition (μ ν : measure α) : Prop :=
(lebesgue_decomposition :
  ∃ (p : measure α × (α → ℝ≥0∞)), measurable p.2 ∧ p.1 ⊥ₘ ν ∧ μ = p.1 + ν.with_density p.2)
```
Since in order to prove the Lebesgue decomposition for $\sigma$-finite measures, 
we will first need to show it for finite measures, this definition allows us 
to reuse the same statement for both cases, avoiding duplicate code. Furthermore, 
as we would like to extract the measure and measurable function from the decomposition, 

> we may define a functions that choose the decomposition if it exists, and are

Reviewer 3: ```suggestion
we may define functions that choose the decomposition if it exists, and are
```
